firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/Sensor.cpp

```diff
@@ -14,65 +14,80 @@ namespace Pufferfish::Driver::Serial::Nonin {
 // Sensor
 
 InitializableState Sensor::setup() {
-  // sensor is connected when we have measurements available
-  switch (measure(measurements_)) {
-    case InitializableState::failed:
+  switch (device_.output(measurements_)) {
+    case PacketStatus::invalid_checksum:
+    case PacketStatus::invalid_header:
       return InitializableState::failed;
-    case InitializableState::ok:
-      sensor_connections_.nonin_connected = true;
-      return InitializableState::ok;
-    case InitializableState::setup:
+    case PacketStatus::waiting:
+    case PacketStatus::frame_loss:
+    case PacketStatus::ok:
+      input_clock(time_.millis());
       break;
   }
 
+  if (wait_time_exceeded()) {
+    return InitializableState::failed;
+  }
+
   return InitializableState::setup;
 }
 
 InitializableState Sensor::output(float &spo2, float &hr) {
```

Comment:
We should return early if last state was not `InitializableState::ok`.

Reply:
Pseudocode:

```
output:
    if we already failed, return failed
    if we're in setup, return setup
    
    if it has been too long since we last received valid data, return failed
    
    // i.e. it hasn't been too long since we last received valid data:
    if we haven't received valid data, return ok early
    
    // i.e. we've received valid data:
    reset the timer
    write the output params
```